File: omninavi/src/main/java/com/omni/omninavi/omninavi/manager/NetworkManager.java
package com.omni.omninavi.omninavi.manager;

import android.content.Context;
import android.net.ConnectivityManager;
import android.provider.Settings;
import android.text.TextUtils;
import android.util.Log;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.TimeoutError;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.gson.Gson;
import com.omni.omninavi.R;
import com.omni.omninavi.omninavi.model.CommonResponse;
import com.omni.omninavi.omninavi.tool.AeSimpleSHA1;
import com.omni.omninavi.omninavi.tool.DialogTools;

import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.security.NoSuchAlgorithmException;
import java.util.Map;

/**
 * Created by wiliiamwang on 31/10/2017.
 */

public class NetworkManager {

    public interface NetworkManagerListener<T> {
        void onSucceed(T object);

        void onFail(VolleyError error, boolean shouldRetry);
    }

    public static final String NETWORKMANAGER_TAG = NetworkManager.class.getSimpleName();
//    public static final String DOMAIN_NAME = "http://nmp.utobonus.com/";
    public static final String DOMAIN_NAME = "https://doit.utobonus.com/";
    public static final String API_RESULT_TRUE = "true";
    public static final String ERROR_MESSAGE_API_TIME_OUT = "Call API timeout";
    public static final int DEFAULT_TIMEOUT = 30000;

    private static NetworkManager mNetworkManager;
    private RequestQueue mRequestQueue;
    private Gson mGson;

    public static NetworkManager getInstance() {
        if (mNetworkManager == null) {
            mNetworkManager = new NetworkManager();
        }
        return mNetworkManager;
    }

    public boolean isNetworkAvailable(Context context) {
        if (context != null) {
            ConnectivityManager manager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            boolean isNetworkEnable = (manager != null &&
                    manager.getActiveNetworkInfo() != null &&
                    manager.getActiveNetworkInfo().isConnectedOrConnecting());

            return isNetworkEnable;
        } else {
            return false;
        }
    }

    public RequestQueue getRequestQueue(Context context) {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(context);
        }
        return mRequestQueue;
    }

    public Gson getGson() {
        if (mGson == null) {
            mGson = new Gson();
        }
        return mGson;
    }

    public String getDeviceId(Context context) {
        return Settings.Secure.getString(context.getContentResolver(), Settings.Secure.ANDROID_ID);
    }

    public <T> void addToRequestQueue(Request<T> req, String tag, Context context) {
        // set the default tag if tag is empty
        req.setTag(TextUtils.isEmpty(tag) ? NETWORKMANAGER_TAG : tag);
        getRequestQueue(context).add(req);
    }

    public <T> void addToRequestQueue(Request<T> req, Context context) {
        req.setTag(NETWORKMANAGER_TAG);
        getRequestQueue(context).add(req);
    }

    public void cancelPendingRequests(Object tag) {
        if (mRequestQueue != null) {
            mRequestQueue.cancelAll(tag);
        }
    }

    private String getMacStr(double currentTimestamp) {
        try {
            return AeSimpleSHA1.SHA1("doitapp://" + currentTimestamp);
        } catch (NoSuchAlgorithmException e) {
            return "";
        } catch (UnsupportedEncodingException e) {
            return "";
        }
    }

    private String addMacInParams(String paramsStr) {
        double currentTimestamp = System.currentTimeMillis() / 1000.0f;
        String mac = getMacStr(currentTimestamp);
        if (!TextUtils.isEmpty(paramsStr)) {
            paramsStr = paramsStr + "&";
        }
        paramsStr = paramsStr + "timestamp=" + currentTimestamp + "&mac=" + mac;

        return paramsStr;
    }

    public <T> void addJsonRequest(final Context context,
                                   int requestMethod,
                                   final String url,
                                   Map<String, String> params,
                                   final Class<T> responseClass,
                                   int timeoutMs,
                                   final NetworkManagerListener<T> listener) {

        if (!isNetworkAvailable(context)) {
            DialogTools.getInstance().dismissProgress(context);
            DialogTools.getInstance().showNoNetworkMessage(context);
            return;
        }

        String paramsString = "";
        if (params != null) {
            for (String key : params.keySet()) {
                if (!TextUtils.isEmpty(paramsString)) {
                    paramsString = paramsString + "&";
                }
                paramsString = paramsString + key + "=" + params.get(key);
            }
        }
        addMacInParams(paramsString);

        final String requestUrl = (TextUtils.isEmpty(paramsString)) ? url : url + "?" + paramsString;
        JsonObjectRequest request = new JsonObjectRequest(requestMethod,
                requestUrl,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        T object = getGson().fromJson(response.toString(), responseClass);
                        listener.onSucceed(object);

                        DialogTools.getInstance().dismissProgress(context);
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        if (error.networkResponse == null && error.getClass().equals(TimeoutError.class)) {
                            error = new VolleyError("Call API timeout");
                        }
                        listener.onFail(error, (TextUtils.isEmpty(error.getMessage()) && error.getCause() == null && error.networkResponse == null));

                        DialogTools.getInstance().dismissProgress(context);
                    }
                }) {

            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                return super.getParams();
            }
        };

        request.setRetryPolicy(new DefaultRetryPolicy(
                timeoutMs,
                DefaultRetryPolicy.DEFAULT_MAX_RETRIES,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));

        addToRequestQueue(request, context.getClass().getSimpleName(), context);
    }

    public <T> void addJsonRequestToCommonObj(Context context,
                                              int requestMethod,
                                              final String url,
                                              Map<String, String> params,
                                              final Class<T[]> responseClass,
                                              final NetworkManagerListener<T[]> listener) {

        addJsonRequestToCommonObj(context, requestMethod, url, params, responseClass, DEFAULT_TIMEOUT, listener);
    }

    public <T> void addJsonRequestToCommonObj(final Context context,
                                              int requestMethod,
                                              final String url,
                                              Map<String, String> params,
                                              final Class<T[]> responseClass,
                                              int timeoutMs,
                                              final NetworkManagerListener<T[]> listener) {

        if (!isNetworkAvailable(context)) {
            DialogTools.getInstance().dismissProgress(context);
            DialogTools.getInstance().showNoNetworkMessage(context);
            return;
        }

        String paramsString = "";
        if (params != null) {
            for (String key : params.keySet()) {
                if (!TextUtils.isEmpty(paramsString)) {
                    paramsString = paramsString + "&";
                }
                paramsString = paramsString + key + "=" + params.get(key);
            }
        }
        addMacInParams(paramsString);

        final String requestUrl = (TextUtils.isEmpty(paramsString)) ? url : url + "?" + paramsString;
        JsonObjectRequest request = new JsonObjectRequest(requestMethod,
                requestUrl,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        CommonResponse commonResponse = getGson().fromJson(response.toString(), CommonResponse.class);
                        if (commonResponse.getResult().equals(API_RESULT_TRUE)) {
                            String json = new Gson().toJson(commonResponse.getData());
                            T[] data = getGson().fromJson(json, responseClass);

                            listener.onSucceed(data);

                            DialogTools.getInstance().dismissProgress(context);
                        } else {
                            DialogTools.getInstance().dismissProgress(context);
                            DialogTools.getInstance().showErrorMessage(context, R.string.error_dialog_title_text_unknown, commonResponse.getErrorMessage());
                        }
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        if (error.networkResponse == null && error.getClass().equals(TimeoutError.class)) {
                            error = new VolleyError("Call API timeout");
                        }
                        listener.onFail(error, (TextUtils.isEmpty(error.getMessage()) && error.getCause() == null && error.networkResponse == null));

                        DialogTools.getInstance().dismissProgress(context);
                    }
                }) {

            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                return super.getParams();
            }
        };

        request.setRetryPolicy(new DefaultRetryPolicy(
                timeoutMs,
                DefaultRetryPolicy.DEFAULT_MAX_RETRIES,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));

        addToRequestQueue(request, context.getClass().getSimpleName(), context);
    }

    public <T> void addPostStringRequest(final Context context,
                                         final String url,
                                         final Map<String, String> params,
                                         final Class<T> responseClass,
                                         int timeoutMs,
                                         final NetworkManagerListener<T> listener) {
        if (!isNetworkAvailable(context)) {
            if (!url.contains("send_user_location") && !url.contains("set_beacon")) {
                DialogTools.getInstance().dismissProgress(context);
                DialogTools.getInstance().showNoNetworkMessage(context);
            }
            return;
        }

        StringRequest request = new StringRequest(Request.Method.POST, url,
                new Response.Listener<String>() {
                    @Override
                    public void onResponse(String response) {
//                        Log.e("@W@", "url : " + url + ", response : " + response);
                        T object = getGson().fromJson(response, responseClass);
                        listener.onSucceed(object);

                        DialogTools.getInstance().dismissProgress(context);
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        if (error.networkResponse != null) {
                            Log.e("NetworkManager", "Error NetworkResponse statusCode === " + error.networkResponse.statusCode);
                        } else {
                            if (error.getClass().equals(TimeoutError.class)) {
                                Log.e("NetworkManager", "*** Error NetworkResponse Timeout, timeMs : " + error.getNetworkTimeMs());
                                error = new VolleyError(ERROR_MESSAGE_API_TIME_OUT);
                            }
                        }
                        listener.onFail(error, (TextUtils.isEmpty(error.getMessage()) && error.getCause() == null && error.networkResponse == null));

                        DialogTools.getInstance().dismissProgress(context);
                    }
                }) {

            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                double currentTimestamp = System.currentTimeMillis() / 1000.0f;
                String mac = getMacStr(currentTimestamp);
                params.put("timestamp", "" + currentTimestamp);
                params.put("mac", mac);

//                Log.e("@W@", "params : " + params.toString());
                return params;
            }
        };

        request.setRetryPolicy(new DefaultRetryPolicy(
                timeoutMs,
                1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));

        addToRequestQueue(request, context.getClass().getSimpleName(), context);
    }
}

File: omninavi/src/main/java/com/omni/navisdk/network/NetworkManager.kt
package com.omni.navisdk.network

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.net.ConnectivityManager
import android.provider.Settings
import android.text.TextUtils
import android.util.Log
import androidx.annotation.DrawableRes
import com.android.volley.*
import com.android.volley.toolbox.ImageLoader
import com.android.volley.toolbox.NetworkImageView
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.gson.Gson
import com.omni.navisdk.R
import com.omni.navisdk.module.CommonArrayResponse
import com.omni.navisdk.module.CommonResponse
import com.omni.navisdk.tool.AeSimpleSHA256
import com.omni.navisdk.tool.DialogTools
import com.omni.navisdk.tool.NaviSDKText
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.UnsupportedEncodingException
import java.security.NoSuchAlgorithmException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class NetworkManager {
    private var mGson: Gson? = null
    private var mRetrofit: Retrofit? = null
    private var mGoogleApiRetrofit: Retrofit? = null
    private var mRequestQueue: RequestQueue? = null

    interface NetworkManagerListener<T> {
        fun onSucceed(response: T?)
        fun onFail(errorMsg: String, shouldRetry: Boolean)
    }

    val retrofit:
            Retrofit?
        get() {
            val okHttpClient = OkHttpClient.Builder()
                    .connectTimeout(TIME_OUT.toLong(), TimeUnit.SECONDS)
                    .writeTimeout(TIME_OUT.toLong(), TimeUnit.SECONDS)
                    .readTimeout(TIME_OUT.toLong(), TimeUnit.SECONDS)
                    .build()
            mRetrofit = Retrofit.Builder()
                    .client(okHttpClient)
                    .baseUrl(DOMAIN_NAME)
                    .callbackExecutor(Executors.newSingleThreadExecutor())
                    .addConverterFactory(GsonConverterFactory.create())
                    .build()
            return mRetrofit
        }

    val googleApiRetrofit: Retrofit?
        get() {
            if (mGoogleApiRetrofit == null) {
                val okHttpClient = OkHttpClient.Builder()
                        .connectTimeout(TIME_OUT.toLong(), TimeUnit.SECONDS)
                        .writeTimeout(TIME_OUT.toLong(), TimeUnit.SECONDS)
                        .readTimeout(TIME_OUT.toLong(), TimeUnit.SECONDS)
                        .build()
                mGoogleApiRetrofit = Retrofit.Builder()
                        .client(okHttpClient)
                        .baseUrl(GOOGLE_API_DOMAIN_NAME)
                        .callbackExecutor(Executors.newSingleThreadExecutor())
                        .addConverterFactory(GsonConverterFactory.create())
                        .build()
            }
            return mGoogleApiRetrofit
        }

    fun getRequestQueue(context: Context?): RequestQueue? {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(context)
        }
        return mRequestQueue
    }

    fun setNetworkImage(context: Context?, networkImageView: NetworkImageView, url: String?) {
        setNetworkImage(context, networkImageView, url, -1, -1)
    }

    fun setNetworkImage(context: Context?, networkImageView: NetworkImageView, url: String?, @DrawableRes errorIconResId: Int) {
        setNetworkImage(context, networkImageView, url, -1, -1)
    }

    fun setNetworkImage(context: Context?, networkImageView: NetworkImageView, url: String?,
                        @DrawableRes defaultIconResId: Int, @DrawableRes errorIconResId: Int) {
        val lruImageCache: LruImageCache = LruImageCache.Companion.instance!!
        val imageLoader = ImageLoader(getRequestQueue(context), lruImageCache)
        networkImageView.setDefaultImageResId(defaultIconResId)
        networkImageView.setErrorImageResId(if (errorIconResId == -1) R.mipmap.syn_poi_information else errorIconResId)
        networkImageView.setImageUrl(url, imageLoader)
    }

    fun isNetworkAvailable(context: Context?): Boolean {
        return if (context != null) {
            val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val wifiNetwork = manager.getNetworkInfo(ConnectivityManager.TYPE_WIFI)
            if (wifiNetwork != null && wifiNetwork.isConnected) {
                return true
            }
            val mobileNetwork = manager.getNetworkInfo(ConnectivityManager.TYPE_MOBILE)
            if (mobileNetwork != null && mobileNetwork.isConnected) {
                return true
            }
            manager.activeNetworkInfo != null && manager.activeNetworkInfo!!.isConnectedOrConnecting
        } else {
            false
        }
    }

    val gson: Gson
        get() {
            if (mGson == null) {
                mGson = Gson()
            }
            return mGson!!
        }

    @SuppressLint("HardwareIds")
    fun getDeviceId(context: Context): String {
        return Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
    }

    private fun checkNetworkStatus(context: Context): Boolean {
        if (!isNetworkAvailable(context)) {
            DialogTools.instance.dismissProgress(context as Activity)
            DialogTools.instance.showNoNetworkMessage(context)
            return false
        }
        return true
    }

    /**
     * Failed response from retrofit.
     */
    private fun sendResponseFailMessage(context: Context,
                                        response: Response<*>,
                                        listener: NetworkManagerListener<*>) {
        val errorMsg = response.message()
        listener.onFail(errorMsg, false)
    }

    /**
     * The error message is from Christine or Mike.
     */
    private fun sendAPIFailMessage(context: Context, errorMsg: String, listener: NetworkManagerListener<*>) {
        listener.onFail(errorMsg, false)
    }

    fun <T> addGetRequestToCommonObj(activity: Activity,
                                     call: Call<CommonResponse>,
                                     responseClass: Class<T>,
                                     listener: NetworkManagerListener<T>) {
        if (!checkNetworkStatus(activity)) {
            return
        }
        call.enqueue(object : Callback<CommonResponse?> {
            override fun onResponse(call: Call<CommonResponse?>, response: Response<CommonResponse?>) {
                if (response.isSuccessful) {
                    if (response.body() == null) {
                        listener.onFail(activity.getString(R.string.error_dialog_title_text_unknown), false)
                    } else {
                        val `object` = gson.fromJson(response.toString(), responseClass)
                        listener.onSucceed(`object`)
                    }
                } else {
                    sendResponseFailMessage(activity, response, listener)
                }
                DialogTools.instance.dismissProgress(activity)
            }

            override fun onFailure(call: Call<CommonResponse?>, t: Throwable) {
                activity.runOnUiThread {
                    listener.onFail(activity.getString(R.string.dialog_message_network_connect_not_good), false)
                    DialogTools.instance.dismissProgress(activity)
                }
            }
        })
    }

    fun <T> addGetRequest(
        activity: Activity,
        call: Call<T>,
        responseClass: Class<T>?,
        listener: NetworkManagerListener<T?>
    ) {
        if (!checkNetworkStatus(activity)) {
            return
        }
        call.enqueue(object : Callback<T?> {
            override fun onResponse(call: Call<T?>, response: Response<T?>) {
                activity.runOnUiThread {
                    if (response.isSuccessful) {
                        if (response.body() == null) {
                            listener.onFail(
                                activity.getString(R.string.error_dialog_title_text_unknown),
                                false
                            )
                        } else {
                            listener.onSucceed(response.body())
                        }
                    } else {
                        sendResponseFailMessage(activity, response, listener)
                    }
                    DialogTools.instance.dismissProgress(activity)
                }
            }

            override fun onFailure(call: Call<T?>, t: Throwable) {
                activity.runOnUiThread { DialogTools.instance.dismissProgress(activity) }
            }
        })
    }

    fun <T> addGetRequestToCommonArrayObj(activity: Activity,
                                          call: Call<CommonArrayResponse>,
                                          responseClass: Class<Array<T>>,
                                          listener: NetworkManagerListener<Array<T>>) {
        if (!checkNetworkStatus(activity)) {
            return
        }
        call.enqueue(object : Callback<CommonArrayResponse?> {
            override fun onResponse(call: Call<CommonArrayResponse?>, response: Response<CommonArrayResponse?>) {
                if (response.isSuccessful) {
                    if (response.body() == null) {
                        Log.e(NaviSDKText.LOG_TAG, "#11")
                        listener.onFail(activity.getString(R.string.error_dialog_title_text_unknown), false)
                    } else {
                        if (response.body()!!.result == API_RESULT_TRUE) {
                            val json = gson.toJson(response.body()!!.data)
                            val data: Array<T>? = gson.fromJson(json, responseClass)
                            if (data != null) {
                                listener.onSucceed(data)
                            }
                        } else {
                            sendAPIFailMessage(activity, response.body()!!.errorMessage, listener)
                        }
                    }
                } else {
                    sendResponseFailMessage(activity, response, listener)
                }
                DialogTools.instance.dismissProgress(activity)
            }

            override fun onFailure(call: Call<CommonArrayResponse?>, t: Throwable) {
                activity.runOnUiThread {
                    listener.onFail(activity.getString(R.string.dialog_message_network_connect_not_good), false)
                    DialogTools.instance.dismissProgress(activity)
                }
            }
        })
    }

    fun <T> addPostRequest(activity: Activity,
                           call: Call<T>,
                           responseClass: Class<T>?,
                           listener: NetworkManagerListener<T?>) {
        addPostRequest(activity, false, call, responseClass, listener)
    }

    fun <T> addPostRequest(activity: Activity,
                           shouldDismissProgress: Boolean,
                           call: Call<T>,
                           responseClass: Class<T>?,
                           listener: NetworkManagerListener<T?>) {
        if (!checkNetworkStatus(activity)) {
            return
        }
        call.enqueue(object : Callback<T?> {
            override fun onResponse(call: Call<T?>, response: Response<T?>) {
                if (response.isSuccessful) {
                    if (response.body() == null) {
                        listener.onFail(activity.getString(R.string.error_dialog_title_text_unknown), false)
                    } else {
                        listener.onSucceed(response.body())
                    }
                } else {
                    sendResponseFailMessage(activity, response, listener)
                }
                if (shouldDismissProgress) {
                    DialogTools.instance?.dismissProgress(activity)
                }
            }

            override fun onFailure(call: Call<T?>, t: Throwable) {
                activity.runOnUiThread {
                    listener.onFail(t.message!!, false)
                    if (shouldDismissProgress) {
                        DialogTools.instance?.dismissProgress(activity)
                    }
                }
            }
        })
    }

    fun <T> addPostRequestToCommonObj(activity: Activity,
                                      call: Call<CommonResponse>,
                                      responseClass: Class<T>?,
                                      listener: NetworkManagerListener<T>) {
        if (!checkNetworkStatus(activity)) {
            return
        }
        call.enqueue(object : Callback<CommonResponse?> {
            override fun onResponse(call: Call<CommonResponse?>, response: Response<CommonResponse?>) {
                if (response.isSuccessful) {
                    if (response.body() == null) {
                        listener.onFail(activity.getString(R.string.error_dialog_title_text_unknown), false)
                    } else {
                        if (response.body()!!.result == API_RESULT_TRUE) {
                            if (response.body()!!.data.toString().isNotEmpty()) {
                                val json = gson.toJson(response.body()!!.data)
                                val data = gson.fromJson(json, responseClass)
                                listener.onSucceed(data)
                            }
                        } else {
                            sendAPIFailMessage(activity, response.body()!!.errorMessage, listener)
                        }
                    }
                } else {
                    sendResponseFailMessage(activity, response, listener)
                }
                DialogTools.instance.dismissProgress(activity)
            }

            override fun onFailure(call: Call<CommonResponse?>, t: Throwable) {
                activity.runOnUiThread {
                    listener.onFail(activity.getString(R.string.dialog_message_network_connect_not_good), false)
                    DialogTools.instance.dismissProgress(activity)
                }
            }
        })
    }

    fun <T> addPostStringRequest(
        context: Context,
        url: String,
        params: MutableMap<String, String>,
        responseClass: Class<T>?,
        timeoutMs: Int,
        listener: NetworkManagerListener<T>
    ) {
        if (!isNetworkAvailable(context)) {
            DialogTools.instance.dismissProgress(context as Activity)
            DialogTools.instance.showNoNetworkMessage(context)
            return
        }
        val request: StringRequest = object : StringRequest(
            Method.POST, url,
            com.android.volley.Response.Listener { response ->
                Log.e("ASD", "url : $url, response : $response")
                val `object`: T = mGson!!.fromJson(response, responseClass)
                listener.onSucceed(`object`)
                DialogTools.instance.dismissProgress(context as Activity)
            },
            com.android.volley.Response.ErrorListener { error ->
                if (error.networkResponse != null) {
                    Log.e(
                        "@W@",
                        "Error NetworkResponse statusCode === " + error.networkResponse.statusCode
                    )
                } else {
                    if (error.javaClass == TimeoutError::class.java) {
                        Log.e(
                            "@W@",
                            "*** Error NetworkResponse Timeout, timeMs : " + error.networkTimeMs
                        )
                    }
                }
                listener.onFail(
                    error.toString(),
                    TextUtils.isEmpty(error.message) && error.cause == null && error.networkResponse == null
                )
                DialogTools.instance.dismissProgress(context as Activity)
            }) {
            @Throws(AuthFailureError::class)
            override fun getParams(): Map<String, String> {
                val currentTimestamp = (System.currentTimeMillis() / 1000.0f).toDouble()
                val mac = getMacStr(currentTimestamp)
                params["timestamp"] = "" + currentTimestamp
                params["mac"] = mac!!
                Log.e("@W@", "params : $params")
                return params
            }
        }
        request.retryPolicy = DefaultRetryPolicy(
            timeoutMs,
            1,
            DefaultRetryPolicy.DEFAULT_BACKOFF_MULT
        )
        addToRequestQueue<String>(request, context.javaClass.simpleName, context)
    }

    fun <T> addToRequestQueue(req: Request<T>, tag: String?, context: Context?) {
        // set the default tag if tag is empty
        req.tag = if (TextUtils.isEmpty(tag)) NetworkManager.NETWORKMANAGER_TAG else tag
        getRequestQueue(context)!!.add(req)
    }

    fun <T> addPostRequestToCommonArrayObj(activity: Activity,
                                           call: Call<CommonArrayResponse>,
                                           responseClass: Class<Array<T>>,
                                           listener: NetworkManagerListener<Array<T>>) {
        addPostRequestToCommonArrayObj(activity, false, call, responseClass, listener)
    }

    fun <T> addPostRequestToCommonArrayObj(activity: Activity,
                                           shouldDismissProgress: Boolean,
                                           call: Call<CommonArrayResponse>,
                                           responseClass: Class<Array<T>>,
                                           listener: NetworkManagerListener<Array<T>>) {
        if (!checkNetworkStatus(activity)) {
            return
        }
        Log.e(NaviSDKText.LOG_TAG, "addPostRequestToCommonArrayObj")
        call.enqueue(object : Callback<CommonArrayResponse?> {
            override fun onResponse(call: Call<CommonArrayResponse?>, response: Response<CommonArrayResponse?>) {
                if (response.isSuccessful) {
                    if (response.body() == null) {
                        Log.e(NaviSDKText.LOG_TAG, "#1")
                        listener.onFail(activity.getString(R.string.error_dialog_title_text_unknown), false)
                    } else {
                        if (response.body()!!.result == API_RESULT_TRUE) {
                            val json = gson.toJson(response.body()!!.data)
                            val data: Array<T>? = gson.fromJson(json, responseClass)
                            if (data != null) {
                                listener.onSucceed(data)
                            }
                        } else {
                            Log.e(NaviSDKText.LOG_TAG, "#2")
                            sendAPIFailMessage(activity, response.body()!!.errorMessage, listener)
                        }
                    }
                } else {
                    Log.e(NaviSDKText.LOG_TAG, "#3 response : $response")
                    sendResponseFailMessage(activity, response, listener)
                }
                if (shouldDismissProgress) {
                    DialogTools.instance.dismissProgress(activity)
                }
            }

            override fun onFailure(call: Call<CommonArrayResponse?>, t: Throwable) {
                activity.runOnUiThread {
                    Log.e(NaviSDKText.LOG_TAG, "#4 localized msg : " + t.localizedMessage + ", msg : " + t.message + ", cause : " + t.cause)
                    listener.onFail(activity.getString(R.string.dialog_message_network_connect_not_good), false)
                    if (shouldDismissProgress) {
                        DialogTools.instance.dismissProgress(activity)
                    }
                    Log.e(NaviSDKText.LOG_TAG, "#5")
                }
            }
        })
    }

    fun getMacStr(currentTimestamp: Double): String? {
        return try {
            AeSimpleSHA256.SHA256(ENCRYPT_KEY + currentTimestamp)
        } catch (e: NoSuchAlgorithmException) {
            ""
        } catch (e: UnsupportedEncodingException) {
            ""
        }
    }

    companion object {
        const val DOMAIN_NAME = "https://navi.taipei/"
        var ENCRYPT_KEY = "doitapp://"
        val NETWORKMANAGER_TAG = NetworkManager::class.java.simpleName
        const val GOOGLE_API_DOMAIN_NAME = "https://maps.googleapis.com/"
        const val API_RESULT_TRUE = "true"
        const val TIME_OUT = 120
        private var sNetworkManager: NetworkManager? = null
        val instance: NetworkManager
            get() {
                if (sNetworkManager == null) {
                    sNetworkManager = NetworkManager()
                }
                return sNetworkManager!!
            }
    }
}